### ISSUE TITLE
Unable to resolve type of if-expr when both branches are loop expressions

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#24402]
```chapel
proc test1() {
  var n = 0;
  var arr = if n then for i in 0..3 do i else for i in 4..7 do i;
  writeln(arr);
}

proc main() {
  test1();
}
```

```
TernaryExpressionBugTwoIters.chpl:1: In function 'test1':
TernaryExpressionBugTwoIters.chpl:3: error: Unable to resolve type of if-expression
note: 'then' branch returns type "_ir_chpl__loopexpr_iter10"
note: 'else' branch returns type "_ir_chpl__loopexpr_iter11"
```

**Comment by bradcray** [comment]
@dlongnecke-cray : I have a feeling this issue has the same root cause as https://github.com/chapel-lang/chapel/issues/24108.  I'd propose we fix this in dyno rather than in the production compiler unless you/someone see(s) an easy path forward.

**Comment by dlongnecke-cray** [comment]
Hi Brad, could you go into more detail here about what you see as the challenges we face solving this issue?

I think I see how this and #24108 might be related - in that it's difficult for us to know with the current compiler whether or not we want to unify the type to an array or an iterator?

What's unclear to me are what we expect the semantics to be. If it's just "promote both branches to an array", then I can probably code that pretty quickly. If it's "promote to an array if we detect the if-expression is used to initialize a variable", that's probably doable too.

I had an idea that we could probably still unify the type to an iterator as well - by creating a third iterator which is the union of these two other iterators (or failing to do so if the types don't match). But @vasslitvinov shot that down because, as I understand it, we want each iterator to have a 1-to-1 mapping with the source code.

**Comment by bradcray** [comment]
To me, both this issue and #24108 are related because (I think) they are both about unifying types across multiple branches of control flow to determine whether they're the same or not.  My instinct would be that we'd want to unify both to an iterator to avoid manifesting an array before/unless it is strictly necessary (following the philosophy of "Chapel tries not to inject arrays unless you ask us to").

(That said, I'm realizing there's also a difference because on #24108, my workaround was to use a conditional expression like this one...  Did that one work where this one doesn't simply by virtue of the iterator being named and called?)

I imagine the "correct" approach here to be somewhat similar to FCP types in that both `for i in 0..3 do i` and `for i in 4..7 do i` have the same static type (e.g., "invoking a 1D iterator over range and yielding `i.type` elements"); yet, today at least, they have different types according to the compiler.  So that suggests either trying to unify to the same static type and having two instances of it, one for each iterator; or using more of a class hierarchy approach as we (currently) do for FCPs.  The latter may have unfortunate performance consequences (possibly what Vass is worried about); the former seems like it could be doable but a fair amount of work.

(I should also note that I am not aware of users in the field being blocked by either of these issues, though I agree they're dissatisfying from a language usability and orthogonality standpoint).

**Comment by dlongnecke-cray** [comment]
It feels like we ought to be able to detect a subset of cases where it is OK to promote to an array. If our pattern matching here is too loose, that means that some cases where we promote a temporary array too early could slip through the cracks. However, this might be an OK cost if it means we could get more programs to work now that didn't work before?

I'm not sure I understand how dyno would enable us to get this to work any faster than the current compiler would. The dynamic typing aspect like a structural problem with the underlying representation of iterators (which is a massive overarching issue?). So maybe looking for some exceptions in the short-term is worth it?

> So that suggests either trying to unify to the same static type and having two instances of it, one for each iterator

Are you suggesting that in the case of something like:

```chapel
iter foo1(): a { ... }
iter foo2(): b { ... }
iter foo3(): c { ... }
// ...
// Assume a 'd' which can convert from a/b/c:
proc bar(d): int { ... }

bar(if x // if-expr2
  then foo3()
  else if y // if-expr1
    then foo2()
    else foo1());
```

We'd effectively generate:

```chapel
int tmp;
if (x) {
  tmp = bar(foo3());
} else {
  if (y) {
    tmp = bar(foo2());
  } else {
    tmp = bar(foo1());
  }
}
```

If so, that probably has a ton of implications I haven't thought about...

But anyhow, if that's what we'd like to try doing, it would definitely help to wait until `dyno` because then we can easily detect if we're able to do this.

**Comment by dlongnecke-cray** [comment]
> I didn't mean to imply maybe we could get it to work sooner in dyno; more like "if it's not blocking users, maybe we should just wait and solve it there (taking advantage of its better architecture and engineering) rather than creating code in the production compiler that we'd eventually need to reimplement in Dyno anyway."

Gotcha. I was just trying to understand if it was essentially a compiler problem or an iterator problem. Sounds like a better compiler can't hurt, though 😄.

> For this example, I think that, today, we don't make any claims that we can unify types across iterator expressions with different signatures—i.e., I don't think the compiler would look at the three branches of the conditional and think "d is a unifying type for these three types being yielded."

I think I complicated my example too much by adding in the different types. Let's imagine that all the iterators just return `int`. In that case we still have the issue today, because each iterator is intended to represent a different loop expression transformed into an iterator.

> So that suggests either trying to unify to the same static type and having two instances of it

Ahhh. Dang it. I read this completely wrong. I mistakenly extrapolated this into an idea where we have a sort of "phi" node which represents a type that can be one of two different types. And then we do something crazy like cloning code to account for all the different branches.

But what you're actually suggesting is the difference between...

```chapel
iter foo1(): int { ... }
iter foo2(): int { ... }
var x = if n then foo1() else foo2();
```

- A sort of "union iterator" which takes one of two different iterators, both of which return the same type (or subtypes), and a `bool` to decide which one to invoke. Then we generate basically:

```chapel
// This is my idea of a sort of "union iterator", but let
// me know if you had a different idea for this!
iter chpl_if_expr_iter(thenBranch: bool): int {
  if thenBranch {
    for x in foo1() do yield x;
  } else {
    for x in foo2() do yield x;
  }
}
var x = chpl_if_expr_iter(n);
```

- And then some sort of class-based representation like we currently use for classes - or, since an iterator essentially seems like an inverted closure, eventually some sort of closure-like thing.

```chapel
class iter_1: iter__int { iter this() { for x in foo1() do yield x; } }
class iter_2: iter__int { iter this() { for x in foo2() do yield x; } }
var x = if n then new shared iter__1() else new shared iter__2();
```

Seem more like what you had in mind?

I really like the look of the first idea myself but I guess there are some problems with it? However, if we're willing to create an even more dynamic representation in the second idea, is the first really so bad?

**Comment by bradcray** [comment]
> A sort of "union iterator" which takes one of two different iterators, both of which return the same type (or subtypes), and a bool to decide which one to invoke.

Oh, that's actually a very interesting idea that hadn't occurred to me.  In some contexts, this would probably work great.  In others, though, I could imagine it adding a lot of overhead.  For example, I think our zippering optimizations wouldn't fire for this union iterator, if it was zipped with something else within `bar()`, because it doesn't have a simple loop structure.

The class-based approach is much more like what I was envisioning, though I worry a little about whether the presence of classes will also thwart optimizations.  So then I was also wondering whether—for the example in the OP at least—we could do something more primitive.  After all, we're invoking the same iterator on both branches, just with different values.  So unlike your foo*() examples, it's actually more like:

```chapel
var x = if n then foo(range1) else foo(range2);
// or:
// var x = if n then foo(range1) else foo(range2);
// or:
// var x = if n then range1.myIter() else range2.myIter();
```

where `arg1` and `arg2` (or `range1` and `range2`) are the same type, just with different values.  So, if `arg1` and `arg2` can be copied/captured cheaply and safely, we could just store that value into two different instances of the same iterator record object, like:

```chapel
var x = if n then new ir_record(arg1) else new ir_record(arg2);
```

or even:

```chapel
var x = new ir_record(if n then arg1 else arg2);
```

(though I expect the first is easier to implement without losing anything).

If `arg1` and `arg2` can't be copied/captured cheaply, then I'd find myself once again wishing we already had `ref` fields implemented (I think we definitely want them) in which case we could just capture the argument by `ref` rather than value.

**Comment by dlongnecke-cray** [comment]
>  After all, we're invoking the same iterator on both branches, just with different values. So unlike your foo*() examples, it's actually more like...

Are you saying that even though the example complains about two different iterator types, `_ir_chpl__loopexpr_iter10` and `_ir_chpl__loopexpr_iter11`, that structurally those two types are the same type (a record containing a simple range), and their loop body is the same (just forwarding the range values), so we could just construct the same type e.g., `ir_record`? 😮

**Comment by bradcray** [comment]
Yeah, exactly.  Or at least, I can't imagine why that would not be the case (or perhaps "could not be made to be the case").

**Comment by dlongnecke-cray** [comment]
Ok, that's pretty dang awesome.

But if it's the case where we can no longer do that, e.g., user iterators that capture structurally different state, then we're kind of back to the drawing board, aren't we?

I really loved my "union iterator" idea and it's too bad that that's going to thwart optimizations. I can't really visualize the class-based approach well enough in my head to have any sense of how it would impact performance compared to iterators today.

**Comment by bradcray** [comment]
> But if it's the case where we can no longer do that, e.g., user iterators that capture structurally different state, then we're kind of back to the drawing board, aren't we?

Yeah, that's what the second edit to my previous comment was about (which I realize you may have missed since it was an edit).

I could also imagine the language (for some time at least) only supporting "trivially unifiable iterator expressions"—those that invoke the same iterator with the same argument types, say.  Put another way, I think Paul is completely justified in feeling like these expressions should unify here and in #24108.  And while the language would be best if foo1() and foo2() both yielding int could be unified, I think we'd also be on reasonable ground if we were to say "We just don't support that (yet)."
